lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,580 @@
+      // records the Operation that is the owner of the corresponding child
+      // value. The operation will only be used for error reporting.
+      getChildren(visitingNode, inputArg, children);
+      // Now continue the DFS recursively.
+      for (auto childNode : children)
+        if (dfsFromNode(childNode.first, inputArg).failed()) {
+          printError(visitingNode, childNode);
+          return failure();
+        }
+      currentPath.pop_back();
+    }
+    currentPath.pop_back();
+
+    visiting = std::move(tmpVisiting);
+    // Finished discovering all the reachable nodes from node.
+    visited.insert(node);
+    return success();
+  }
+
+  // Local implementation of getFieldRefFromValue, returns the Aggregate type
+  /// Get the FieldRef from a value. Customized implementation of
+  /// getFieldRefFromValue. If `field` is a subfield, then this will return the
+  /// base aggregate Value and the corresponding fieldId. This will return
+  /// nullptr if thie field is a result of a register and can be ignored.
+  FieldRef getFieldRef(Value field) {
+    if (field.isa<BlockArgument>() ||
+        !isa<SubfieldOp, SubaccessOp, SubindexOp>(field.getDefiningOp()))
+      return FieldRef(field, 0);
+    auto iter = leafToBaseValMap.find(field);
+    if (iter == leafToBaseValMap.end())
+      return FieldRef(nullptr, 0);
+    auto baseVal = iter->second;
+    auto storIter = valToStorageLocIdMap.find(baseVal);
+    if (storIter == valToStorageLocIdMap.end())
+      return FieldRef(nullptr, 0);
+    auto baseStorId = storIter->getSecond().front();
+    auto fieldStorIter = valToStorageLocIdMap.find(field);
+    if (fieldStorIter == valToStorageLocIdMap.end())
+      return FieldRef(nullptr, 0);
+    auto fieldStorId = fieldStorIter->getSecond().front();
+
+    return FieldRef(baseVal, fieldStorId - baseStorId);
+  }
+
+  // Add all the Values to `children` that have a valid dataflow from `node`.
+  // The `inputArg` is valid, only if it is one of the reaching definitions to
+  // `node`.
+  void getChildren(Value node, Node &inputArg,
+                   SmallVector<ValueOpPair> &children) {
+
+    // Handle the following cases
+    // Case: Instance port: Add the port fields that have a combinational path
+    // in the referenced module, to the children.
+    //
+    if (!node.isa<BlockArgument>()) {
+      // Any Value that is derived from a Register must be ignored. `skip` is
+      // true, only if this node is the result of some Register.
+      bool skip =
+          TypeSwitch<Operation *, bool>(node.getDefiningOp())
+              .Case<InstanceOp>([&](InstanceOp ins) {
+                size_t portNum = 0;
+                // Get the port index for `node`.
+                for (const auto &p : llvm::enumerate(ins->getOpResults()))
+                  if (p.value() == node) {
+                    portNum = p.index();
+                    break;
+                  }
```

Comment:
possibly we can use `node.cast<mlir::OpResult>().getResultNumber()`

Reply:
Thanks, this is much better.